programs/svm-spoke/src/utils/message_utils.rs

```diff
@@ -0,0 +1,102 @@
+        if account_infos[i].key() != message_account_key {
+            return Err(Error::from(AcrossPlusError::InvalidMessageAccountKey)
+                .with_pubkeys((account_infos[i].key(), message_account_key)));
+        }
+
+        // Writable accounts must be passed first. This enforces the same write permissions as set in the message. Note
+        // that this would fail if any of mutable FillV3Relay / ExecuteV3SlowRelayLeaf accounts are passed as read-only
+        // in the bridged message as the calling client deduplicates the accounts and applies maximum required
+        // privileges. Though it is unlikely that any practical application would require this.
+        // We also explicitly disable all signer privileges for all the accounts to protect the relayer from being
+        // drained of funds in the inner instructions.
+        match i < message_accounts_len - (message.read_only_len as usize) {
+            true => {
+                if !account_infos[i].is_writable {
+                    return Err(Error::from(AcrossPlusError::NotWritableMessageAccountKey)
+                        .with_account_name(format!("{}", message_account_key)));
+                }
+                accounts.push(AccountMeta::new(message_account_key, false));
+            }
+            false => {
+                if account_infos[i].is_writable {
+                    return Err(Error::from(AcrossPlusError::NotReadOnlyMessageAccountKey)
+                        .with_account_name(format!("{}", message_account_key)));
+                }
+                accounts.push(AccountMeta::new_readonly(message_account_key, false));
+            }
+        }
+    }
+
+    // Transfer value amount from the relayer to the first account in the message accounts.
+    // Note that the depositor is responsible to make sure that after invoking the handler the recipient account will
+    // not hold any balance that is below its rent-exempt threshold, otherwise the fill would fail.
+    if message.value_amount > 0 {
+        let recipient_account = account_infos.get(0).ok_or(AcrossPlusError::MissingValueRecipientKey)?;
+        let transfer_ix = system_instruction::transfer(&relayer.key(), &recipient_account.key(), message.value_amount);
+        invoke(
+            &transfer_ix,
+            &[relayer.to_account_info(), recipient_account.to_account_info()],
+        )?;
+    }
+
+    // The data will hold the handler ix discriminator and raw handler message bytes (including 4 bytes for the length).
+    let mut data = Vec::with_capacity(DISCRIMINATOR_SIZE + 4 + message.handler_message.len());
+    data.extend_from_slice(&HANDLE_V3_ACROSS_MESSAGE_DISCRIMINATOR);
+    AnchorSerialize::serialize(&message.handler_message, &mut data)?;
+
+    let instruction = Instruction {
+        program_id: message.handler,
+        accounts,
+        data,
+    };
+
+    // TODO: consider if the message handler requires signed invocation.
```

Comment:
This mostly depends on whether the handler needs to retain any privileges on behalf of the user. E.g. if the handler performs a swap and forwards swapped-in tokens to user's ATA then its not important. On the other hand, if the handler deposits funds in a money market that does not allow depositing on behalf of end user then the handler would need to retain rights to withdraw the funds where such call must be authenticated by spoke pool PDA that is derived from depositing user address and origin chain.

Reply:
right. main issue being how much size we have left in the tx. I think we can circle back to this once we've totally validated the upper bound of the tx sizing here and that it covers enough basic usecases.

Reply:
for example this tx that we did from the dev wallet: https://solscan.io/tx/5swoRhHcqpxRXRhyyaCRWqTGDcNuQ8avZXptLomyUFPEPKjQceiMfSPdypA9pcdZQu3inGHmhoZmBWdEbZwGRZYC

this is a deposit into a money market protocol from the perspective of the user. this would not work in the current pattern given this restriction. it could be possible to re-structure this tx such that the relayer deposits and then transfers ownership of the position to the user but this seems a fair bit more complex and requires this kind of transfer logic within the target protocol. 

in general though not having authentication might be ok? we need to think about the exact user flow that we want to accommodate for across+.

Reply:
ok after discussion this more we've decided that this is infact _not_ a requirement as the filler on evm is not permissioned (no preservation of msg.sender on evm) so this is fine to to have on svm.